src/diagnostics/MOM_harmonic_analysis.F90

```diff
@@ -0,0 +1,411 @@
+    call MOM_error(NOTE, trim(mesg))
+
+    ! Get the lower and upper bounds of input data
+    ha1%is = LBOUND(data,1) ; is = ha1%is
+    ha1%ie = UBOUND(data,1) ; ie = ha1%ie
+    ha1%js = LBOUND(data,2) ; js = ha1%js
+    ha1%je = UBOUND(data,2) ; je = ha1%je
+
+    allocate(ha1%ref(is:ie,js:je), source=0.0)
+    allocate(ha1%FtF(2*nc+1,2*nc+1), source=0.0)
+    allocate(ha1%FtSSH(is:ie,js:je,2*nc+1), source=0.0)
+
+    do j=js,je ; do i=is,ie
+      ha1%ref(i,j) = data(i,j)
+    enddo ; enddo
+  else
+    dt = now - ha1%old_time
+    is = ha1%is ; ie = ha1%ie ; js = ha1%js ; je = ha1%je
+  endif
+
+  ha1%old_time = now                        !< Keep track of time so we know when Time approaches CS%time_end
+
+  ! Accumulate FtF and FtSSH
+  ha1%FtF(1,1) = ha1%FtF(1,1) + 1.0         !< For the zero frequency
+  do c=1,nc
+    icos = 2*c
+    isin = 2*c+1
+    cosomegat = cos(CS%freq(c) * now + CS%phase0(c))
+    sinomegat = sin(CS%freq(c) * now + CS%phase0(c))
+    ha1%FtF(icos,1) = ha1%FtF(icos,1) + cosomegat
+    ha1%FtF(isin,1) = ha1%FtF(isin,1) + sinomegat
+    ha1%FtF(1,icos) = ha1%FtF(icos,1)
+    ha1%FtF(1,isin) = ha1%FtF(isin,1)
+    do cc=c,nc
+      iccos = 2*cc
+      issin = 2*cc+1
+      ccosomegat = cos(CS%freq(cc) * now + CS%phase0(cc))
+      ssinomegat = sin(CS%freq(cc) * now + CS%phase0(cc))
+      ha1%FtF(icos,iccos) = ha1%FtF(icos,iccos) + cosomegat * ccosomegat
+      ha1%FtF(icos,issin) = ha1%FtF(icos,issin) + cosomegat * ssinomegat
+      ha1%FtF(isin,iccos) = ha1%FtF(isin,iccos) + sinomegat * ccosomegat
+      ha1%FtF(isin,issin) = ha1%FtF(isin,issin) + sinomegat * ssinomegat
+    enddo ! cc=c,nc
+    do j=js,je ; do i=is,ie
+      ha1%FtSSH(i,j,1)    = ha1%FtSSH(i,j,1)    + (data(i,j) - ha1%ref(i,j))
+      ha1%FtSSH(i,j,icos) = ha1%FtSSH(i,j,icos) + (data(i,j) - ha1%ref(i,j)) * cosomegat
+      ha1%FtSSH(i,j,isin) = ha1%FtSSH(i,j,isin) + (data(i,j) - ha1%ref(i,j)) * sinomegat
+    enddo ; enddo
+  enddo ! c=1,nc
+
+  ! Compute harmonic constants and write output as Time approaches CS%time_end
+  ! This guarantees that HA_write will be called before Time becomes larger than CS%time_end
+  if (time_type_to_real(CS%time_end - Time) <= dt) then
+    call HA_write(ha1, Time, G, CS)
+
+    write(mesg,*) "HAmod: harmonic analysis done, key = ", trim(ha1%key)
+    call MOM_error(NOTE, trim(mesg))
+  endif
+
+end subroutine HA_accum
+
+!> This subroutine computes the harmonic constants and write output for the current field
+subroutine HA_write(ha1, Time, G, CS)
+  type(HA_type), pointer, intent(in) :: ha1
+  type(time_type),        intent(in) :: Time       !< The current model time
+  type(ocean_grid_type),  intent(in) :: G          !< The ocean's grid structure
+  type(HA_CS),            intent(in) :: CS         !< Control structure of the MOM_harmonic_analysis module
+
+  ! Local variables
+  real, dimension(:,:,:), allocatable :: FtSSHw    !< An array containing the harmonic constants
```

Comment:
What are the units of `FtSSHw`?  If they are arbitrary, please indicate this with a phrase like `in arbitrary units [A]`.